### PR TITLE
[USING] Create Wasabi Setup in Virtual Machines

### PR DESCRIPTION
**Pull request by MaxHillebrand**
This creates a step by step guide on how to compile Wasabi from source in Qubes. This is especially useful for testers of cutting edge developer branches, as it properly separates the mainnet and testnet environment.

This ready for review PR explains how to: Generate Template VM, Install Dependencies, Generate App VM, Compile and Run Wasabi, and Install Package.

### CODE REVIEW

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  "title": "Wasabi Setup on Qubes",
+  "description": "A step by step guide on how to use Wasabi Wallet in Qubes OS. This is the Wasabi documentation, an archive of knowledge about the open-source, non-custodial and privacy-focused Bitcoin wallet for desktop."
+}
+---
+
+# Wasabi Setup on Qubes
+
+[[toc]]
+
+[Qubes](https://qubes-os.org/) is a free and open-source security-oriented operating system meant for single-user desktop computing.
```
Comment (RiccardoMasutti) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
[Qubes](https://www.qubes-os.org/) is a free and open-source security-oriented operating system meant for single-user desktop computing.
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+Qubes OS leverages xen-based virtualization to allow for the creation and management of isolated virtual machines called qubes.
```
Comment (RiccardoMasutti) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
Qubes OS leverages xen-based virtualization to allow for the creation and management of isolated virtual machines (VMs) called qubes.
```

**Comment by MaxHillebrand** [comment]
I was contemplating this too @davterra, though I think it is part of `using-wasabi`, as this is in general a nice idea to have your private keys separated in a VM, even if you are not compiling developer branches, but only using packaged versions. It is definitely edge case, but I think it is good enough to put it in the lowest order of the `Install` section.

**Comment by davterra** [comment]
> I was contemplating this too @davterra, though I think it is part of `using-wasabi`, as this is in general a nice idea to have your private keys separated in a VM, even if you are not compiling developer branches, but only using packaged versions. It is definitely edge case, but I think it is good enough to put it in the lowest order of the `Install` section.

That makes sense. I've been doing the same template/clone idea using Virtualbox. I used to use Qubes, but sadly no longer am able to.

Actually, everything I do on my computer is through Virtualbox. My firewall doesn't allow traffic in or out to the host. All VM's use a bridged connection. (Poor man's Qubes reproduction)

Anyway, Do you think a Virtualbox-oriented page similar to your Qubes page would be of value? I'd be happy to construct it, if so.
While Qubes is more secure in its VM process, there are many more users of Virtualbox.

**Comment by MaxHillebrand** [comment]
Yes @davterra, that would be a nice addition! 
How about we change this chapter to be `Virtual Wasabi` or something, and then have one part being qubes, the other being VirtualBox?

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+[[toc]]
+
+[Qubes](https://www.qubes-os.org/) is a free and open-source security-oriented operating system meant for single-user desktop computing.
+Qubes OS leverages xen-based virtualization to allow for the creation and management of isolated virtual machines (VMs) called qubes.
+This is valuable for users of high security software like Wasabi Wallet, and especially useful for testers of cutting edge developer branches.
+There is a separation of the virtual machines running Wasabi on testnet, and those running on mainnet with valuable private keys.
+If a reviewer compiles a malicious developer branch, then his mainnet private keys are secure in a different qube.
+
+This is a step by step guide on how to properly compile Wasabi Wallet from source in independent qubes.
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
This is a step-by-step guide on how to properly compile Wasabi Wallet from source in independent qubes.
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+```sh
+[user@dom0 ~]$ qvm-clone template-debian-10 template-wasabi
+```
+
+## Install Dependencies
+
+Start the gnome-terminal of `template-wasabi`.
+
+```sh
+[user@dom0 ~]$ qvm-run -a template-wasabi gnome-terminal
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
qvm-run -a template-wasabi gnome-terminal
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+This is a step by step guide on how to properly compile Wasabi Wallet from source in independent qubes.
+For tutorials on how to [download](https://www.qubes-os.org/downloads/) and [install](https://www.qubes-os.org/doc/installation-guide/) Qubes, please see their exhaustive documentation.
+
+## Generate Template VM
+
+The dependencies to [compile Wasabi from source](/using-wasabi/BuildSource.md) will be downloaded and installed in a new template VM, which is generated by cloning the Debian template from the `dom0` terminal.
+
+```sh
+[user@dom0 ~]$ qvm-clone template-debian-10 template-wasabi
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
qvm-clone template-debian-10 template-wasabi
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+Start the gnome-terminal of `template-wasabi`.
+
+```sh
+[user@dom0 ~]$ qvm-run -a template-wasabi gnome-terminal
+```
+
+Update the `template-wasabi`.
+
+```sh
+[user@template-wasabi ~]$ sudo apt update
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
sudo apt update
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+[user@template-wasabi ~]$ sudo apt upgrade
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
sudo apt upgrade
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+[user@template-wasabi ~]$ sudo apt dist-upgrade
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
sudo apt dist-upgrade
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+```
+
+Install [.NET Core ${dotnetVersion} SDK](https://www.microsoft.com/net/download) for "Building Apps" in `template-wasabi`.
+
+:::tip Optional for privacy
+You can disable .NET's telemetry, which is sending some usage information to Microsoft:
+
+```sh
+[user@template-wasabi ~]$ export DOTNET_CLI_TELEMETRY_OPTOUT=1
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
export DOTNET_CLI_TELEMETRY_OPTOUT=1
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+```
+:::
+
+If you need to update .NET Core, then do it in this VM.
+Now shut down `template-wasabi`.
+
+```sh
+[user@dom0 ~]$ qvm-shutdown template-wasabi
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
qvm-shutdown template-wasabi
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+## Generate App VM
+
+Create a new App VM `source-wasabi` which is based on `template-wasabi`.
+This qube will be used to download, compile and run the Wasabi source code.
+This way you can create multiple App VMs for different testing purposes.
+All these App VMs will be updated when you update `template-wasabi`.
+
+```sh
+[user@dom0 ~]$ qvm-create source-wasabi --template template-wasabi --label red
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
qvm-create source-wasabi --template template-wasabi --label red
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+```
+
+## Compile and Run Wasabi
+
+Start the gnome-terminal of `source-wasabi`.
+
+```sh
+[user@dom0 ~]$ qvm-run -a source-wasabi gnome-terminal
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
qvm-run -a source-wasabi gnome-terminal
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+```
+
+Clone & build Wasabi
+
+```sh
+[user@source-wasabi ~]$ git clone https://github.com/zkSNACKs/WalletWasabi.git
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
git clone https://github.com/zkSNACKs/WalletWasabi.git
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+[user@source-wasabi ~]$ cd WalletWasabi/WalletWasabi.Gui
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
cd WalletWasabi/WalletWasabi.Gui
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+[user@source-wasabi ~]$ dotnet build
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
dotnet build
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+```
+
+Add the development branches to your git.
+
+```sh
+[user@source-wasabi ~]$ git remote add nopara73 https://github.com/nopara73/WalletWasabi
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
git remote add nopara73 https://github.com/nopara73/WalletWasabi
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+[user@source-wasabi ~]$ git fetch nopara73
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
git fetch nopara73
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+[user@source-wasabi ~]$ git checkout <development branch>
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
git checkout <development branch>
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+```
+
+Pull the latest commits and compile from source.
+
+```sh
+[user@source-wasabi ~]$ git pull
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
git pull
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+[user@source-wasabi ~]$ dotnet run
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
dotnet run
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+```
+
+## Install Package in App VM
+
+Alternatively, you can also install the package in a new App VM, this is a stable version suitable to use on mainnet and is separated from the development App VM.
+There are no additional dependencies required, so the App VM can be based on `template-debian-10` directly.
+
+```sh
+[user@dom0 ~]$ qvm-create package-wasabi --template template-debian-10 --label red
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
qvm-create package-wasabi --template template-debian-10 --label red
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+[user@dom0 ~]$ qvm-run -a package-wasabi gnome-terminal
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
qvm-run -a package-wasabi gnome-terminal
```

File: docs/using-wasabi/WasabiSetupQubes.md [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+```
+
+[Download, verify and install](/using-wasabi/InstallPackage.md#debian-and-ubuntu) the latest `Wasabi-${currentVersion}.deb` package in `package-wasabi`, then start Wasabi.
+
+```sh
+[user@package-wasabi ~]$ wassabee
```
Comment (yahiheb) [inline docs/using-wasabi/WasabiSetupQubes.md]:
```suggestion
wassabee
```

**Comment by MaxHillebrand** [comment]
@yahiheb I put the `[user@package-wasabi ~]$` there specifically, because here we are working with three different terminals in three different virtual machines. The VMs are created and started in `dom0`, then the dependencies installed in `template-wasabi` and ultimately the source code compiled in `source-wasabi`. I was not sure if it is sufficient to explain this in the text above the code boxes. Thus I chose to put in the whole terminal content, including the VM environment that started the terminal. 

This is why I NACK your suggestions. What do you say?

**Comment by MaxHillebrand** [comment]
@davterra I changed the structure to be `/using-wasabi/WasabiSetupVM.md`, with a first level header `# Wasabi Setup in Virtual Machines` and a second level header `## Qubes` then the previously written in third level headers `###`.

I suggest now to merge this PR, and you opening a new PR that inserts your VM setup environment, starting with a second level header `## VirtualBox`

**Comment by yahiheb** [comment]
> @yahiheb I put the [user@package-wasabi ~]$ there specifically to avoid confusion, because here we are working with four different terminals in four different virtual machines. The VMs are created and started in dom0, then the dependencies installed in template-wasabi and ultimately the source code compiled in source-wasabi. I was not sure if it is sufficient to explain this in the text above the code boxes. Thus I chose to put in the whole terminal content, including the VM environment that started the terminal.
> 
> This is why I NACK your suggestions. What do you say?

That makes sense.

**Comment by davterra** [comment]
> Yes @davterra, that would be a nice addition!
> How about we change this chapter to be `Virtual Wasabi` or something, and then have one part being qubes, the other being VirtualBox?



> @davterra I changed the structure to be `/using-wasabi/WasabiSetupVM.md`, with a first level header `# Wasabi Setup in Virtual Machines` and a second level header `## Qubes` then the previously written in third level headers `###`.
> 
> I suggest now to merge this PR, and you opening a new PR that inserts your VM setup environment, starting with a second level header `## VirtualBox`

Yes, all good!

**Pull request merged**